### PR TITLE
Parsing of manifest.json as well as better yaml parsing support / alias support

### PR DESCRIPTION
**Pull request by z3z1ma**
Picking up the ball where fernandobrito left off. I have a major enhancement (one of two in the works). Lots of cleaned up code, bug fixes from the other fork, personally tested on my own production replicates and internally at my company, etc. There may be some more changes/commits I add (to readme mostly) but I just wanted to post it to get your eyes and awareness. The other PR will involve auto documenting all exposures revealed in metabase through questions / dashboards and constructing a YML that allows us to sync a DAG with metabase fully including click throughs from dbt docs exposures directly to dashboards.

### CODE REVIEW

**Comment by fernandobrito** [comment]
Hi @z3z1ma, great that you picked up the work of merging my branch, thanks! I was away for a month on vacation and now that I'm back there are so many other priorities in my company that I'm not sure when I would be able to come back to my contribution here. Let me know if you or @gouline have any questions on any of the code that I wrote.

**Comment by fernandobrito** [comment]
On a side note, I'm very curious about your next PR where you mention "involve auto documenting all exposures revealed in metabase". Over the next 2 weeks, I will open-source an internal Python project from my company on doing exactly the same, but for Tableau dashboards. I use the Tableau Metadata API to extract custom SQL and dependencies from reports to tables in the database, then I read all dbt models and sources from the `manifest.json` and try to match Tableau dashboards to dbt models/sources. After that, I write those dependencies back in the `manifest.json` as exposures.

I'm naming the Python package "exposures-crawler" because I wanted at some point in the future to include Metabase and other clients as well. I will write here once the package is public, but I already wrote a tiny bit about it on the dbt Slack community: https://getdbt.slack.com/archives/C0VLNUUTZ/p1617012523148700.

> constructing a YML that allows us to sync a DAG with metabase fully including click throughs from dbt docs exposures directly to dashboards.

That last part I didn't really understand.

File: README.rst
Comment (fernandobrito) [inline README.rst:17]: @z3z1ma were you. using 3.6 locally for developing this? Just double-checking, as I was not and there's some chance that some of the syntax we use in the type hints that we introduce require higher versions.

File: dbtmetabase/__init__.py
Comment (fernandobrito) [inline dbtmetabase/__init__.py:37]: Just an extra reminder that although we keep the `dbt_path` flow (folder) for backward compatibility, renaming this argument will be a breaking change to existing users.

File: dbtmetabase/models/metabase.py
Comment (fernandobrito) [inline dbtmetabase/models/metabase.py:9]: Maybe this was overkill on my part, but I had just finished reading some "advanced" Python books and wanted to put some concepts into practice 😄 . It's a descriptor (https://realpython.com/python-descriptors/) to be used in class attributes that automatically convert empty strings to None.

File: dbtmetabase/parsers/dbt_folder.py
Comment (fernandobrito) [inline dbtmetabase/parsers/dbt_folder.py:30]: I didn't really get this part. Isn't `expanduser()` safe to be used in the entire string?
```
>>> os.path.expanduser('~/needs/expansion')
'/Users/brito/needs/expansion

>>> os.path.expanduser('/does/not/need/expansion')
'/does/not/need/expansion'
```

I think there's no need to check if there's `~` in the string, and you can apply in the entire path (no need to break it into two).

File: dbtmetabase/parsers/dbt_folder.py
Comment (fernandobrito) [inline dbtmetabase/parsers/dbt_folder.py:37]: Not a big deal, but I think it's a best practice that type hinting on arguments should be as permissive as possible. For example: Instead of requiring consumers to pass a list, I think any `Iterable` would do here. https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html#standard-duck-types. But then, return types should be as specific as possible.

I think we are also getting a bit inconsistent here using `list` and `List` as type hints. If I'm not mistaken, we can use `list` everywhere as long as we don't mind making this Python 3.9+ only. https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html#standard-duck-types

File: dbtmetabase/parsers/dbt_folder.py
Comment (fernandobrito) [inline dbtmetabase/parsers/dbt_folder.py:121]: Another example of the above. `Mapping` as a type hint could be preferred over `dict`.

**Review by gouline — COMMENTED** [review]
As a first pass, I'll let you two discuss amongst yourselves any existing comments and rebase the fork properly so that diffs are based on the existing state of the master branch (and not weeks back).

Also, this marks a significant increase in code, so before it gets merged, we need unit tests. Happy to instrument execution in GitHub Actions myself, but the author of the code will be responsible for the actual test cases.

File: README.rst
Comment (z3z1ma) [reply README.rst:17]: I was not using 3.6 but I am going to fallback and adjust the type hinting as per your suggestions to maintain wider compatibility for users @fernandobrito .

File: dbtmetabase/__init__.py
Comment (z3z1ma) [reply dbtmetabase/__init__.py:37]: Good call. Ill emphasize the distinction between the 2 database args in the docstring and help text and revert the name to ensure seamless compatibility.

**Comment by z3z1ma** [comment]
> > constructing a YML that allows us to sync a DAG with metabase fully including click throughs from dbt docs exposures directly to dashboards.
> 
> That last part I didn't really understand.

Hey @fernandobrito

To clarify, I mean the dbt docs can also serve as a index which links users directly to the exposures that have been parsed. The reverse essentially of what you did adding dbt doc url appending to metabase decription. There is actually a button (see below screencap) whose link is derived from the exposures yaml. 

image


> On a side note, I'm very curious about your next PR where you mention "involve auto documenting all exposures revealed in metabase". Over the next 2 weeks, I will open-source an internal Python project from my company on doing exactly the same, but for Tableau dashboards. I use the Tableau Metadata API to extract custom SQL and dependencies from reports to tables in the database, then I read all dbt models and sources from the `manifest.json` and try to match Tableau dashboards to dbt models/sources. After that, I write those dependencies back in the `manifest.json` as exposures.
> 

Are you using sqlparse? I've used just regex myself to extract (table level) exposures from native sql questions in metabase with pretty good success actually. I keep the native sql use in metabase low opting instead for better models which allow users to answer questions using just the GUI and tools. In a previous go around there was a lot of native SQL I allowed myself and my users in metabase, but it is definitely not ideal and dbt has been a gamechanger there in how quickly I was able to simplify our extraction of value from the underlying data.

> I'm naming the Python package "exposures-crawler" because I wanted at some point in the future to include Metabase and other clients as well. I will write here once the package is public, but I already wrote a tiny bit about it on the dbt Slack community: https://getdbt.slack.com/archives/C0VLNUUTZ/p1617012523148700. If according to the author of dbt-metabase, your exposures automation is out of scope, we could discuss teaming up together and merging our 2 projects.
> 

I am curious on yours too. My implementation involves constructing a yaml (we can call it `metabase-exposures.yml` in root model dir for example) over manipulating `manifest.json` (though I will certainly read from the manifest). I am taking this approach because it doesn't involve continuously executing a script or an additional step in CI/CD every single time dbt is invoked and the manifest.json is reconstructed. Instead a yaml is very readable, introspectable, and all subsequent dbt runs will add exposures to manifest implicitly without a post-run dependency. It also lets us revision and commit the yaml to source control too for what that's worth. I'd guarantee there an intersect of our interests and implementations that would be beneficial though so keep me posted. The end result of both of our methods is (should be) a manifest/doc site with perfectly documented exposures.

File: README.rst
Comment (z3z1ma) [reply README.rst:147]: Hey @gouline 

I actually took the latest master branch and personally went through manually integrating changes from fernandobritos branch, so we should have he best of both worlds. Thanks for the catch on the above though, no reason to take that out. 

> EDIT: Unit testing structure, linting and type checking now in place. Please rebase against master and make sure all of that passes for this pull request before writing tests.
>

Thanks a bunch for the added tests. I will rebase and get tests completed.

File: dbtmetabase/parsers/dbt_folder.py
Comment (z3z1ma) [reply dbtmetabase/parsers/dbt_folder.py:33]: Updated type hints to be compatible with pre 3.9 releases of python. Made them all consistent too. Thanks for the link, I'll be putting that principle into practice more in my own coding- thanks!

**Comment by gouline** [comment]
@z3z1ma Just as an FYI, you can run all validation locally with 'make lint type test' without having to push each commit.

**Comment by z3z1ma** [comment]
@gouline 

Thanks for the heads up and sorry for blowing up your inbox then lol. Appreciate it.

**Comment by z3z1ma** [comment]
@gouline 

Let me know next steps,

@fernandobrito & @gouline 

I have the next PR about ready once and if the master is updated. See below examples. All Analyses, Dashboards, and even native sql exposures in Metabase are automatically documented and compiled into a dbt yml. This includes creator/owner of analysis, click through from exposure documentation to exposure, etc. Getting a lot of positive feedback from dbt community on this one. 

image

image

image

**Comment by gouline** [comment]
@z3z1ma Remaining step for this PR is tests. Then @fernandobrito and I can both review the whole thing and I'll merge it when everyone's happy.

**Comment by z3z1ma** [comment]
Sounds good. So with the structure you have in place it's just a matter of providing mock data right? Since it passed everything else? Let me know if there are any other considerations and thanks for your guidance. I appreciate you. If I'm on the right track I'll knock it out tonight.

**Comment by gouline** [comment]
Mock data for each reader and assertions for all the edge cases. The objective isn't as much to test dbt schema and manifest files (they do that themselves) but to ensure that your code is handling all the use cases properly and when future PRs come through and change that code, we're not breaking anything that you intended.

**Comment by z3z1ma** [comment]
@gouline 

First pass unit tests complete. Think they look really good. Uniform in expected output. 
The Metabase client test is going to be a little more involved (though I haven't changed it, functionally). Wondering if you have any ideas there to make the test simpler- otherwise my only additional comment is I have a PR I want to push down the pipeline after this which does add a function to Metabase client (exposure extraction) and uses existing function of our two parsers to cross reference models generating YAML for DAG. I wouldn't mind diving more deeply into Metabase client tests in that PR. 
Either was is okay, just lmk what you think direction-wise. Thanks again!

**Comment by gouline** [comment]
Yeah, don't bother with Metabase client tests in this PR.

If this is ready for review, @fernandobrito can you please review the changes first, since it's based on your original code? When done, I'll do a review too.

File: dbtmetabase/__init__.py
Comment (fernandobrito) [inline dbtmetabase/__init__.py:18]: Is it intentional that the default value of 30 was removed for `sync_timeout`?

File: dbtmetabase/__init__.py
Comment (fernandobrito) [inline dbtmetabase/__init__.py:143]: Is the change from `True` to `False` intentional?

File: dbtmetabase/__init__.py
Comment (fernandobrito) [inline dbtmetabase/__init__.py:175]: Also here the default changed from `True` to `False`, or is it me misunderstanding `action="store_true"`? 😅

File: dbtmetabase/__init__.py
Comment (fernandobrito) [inline dbtmetabase/__init__.py:213]: Not a big deal, but some strings in `help` end with a period (`.`), and others don't. Not sure which one to use, though. Maybe the author of the library has a preferred format?

File: dbtmetabase/metabase.py
Comment (fernandobrito) [inline dbtmetabase/metabase.py:22]: Isn't `verify` a `str` at this point?

File: dbtmetabase/metabase.py
Comment (fernandobrito) [inline dbtmetabase/metabase.py:124]: Note to the final reviewer: `lookup_key` was introduced instead of just `model_name` because in my setup I have tables with the same name in different schemas, so just using the table name was not enough.

File: tests/LICENSE
Comment (fernandobrito) [inline tests/LICENSE:1]: Do the tests have a different license than the main project?

File: tests/README.md
Comment (fernandobrito) [inline tests/README.md:1]: I found it a bit confusing that the sample dbt project is directly in the `tests` root folder. I would find it a bit more tidy if it were on a folder like `fixtures/sample_project` or something similar.

File: tests/test_dbt_folder_reader.py
Comment (fernandobrito) [inline tests/test_dbt_folder_reader.py:257]: Remove this one?

**Comment by fernandobrito** [comment]
Ehehe, it ended up being a quite long PR. I didn't have time to actually clone this locally and try to run it against the real dbt and Metabase instances where I work, but I went through the code and added a few comments/questions.

Something that would be nice would be to have some sort of plan to deprecate the YAML parsing feature in a couple of releases. That could help reducing the codebase.

Thanks again @z3z1ma for picking this up. Next week I am busy, but I would also be available to address comments or write more tests afterward if needed.

File: tests/README.md
Comment (z3z1ma) [reply tests/README.md]: Agreed, moved to a fixtures folder in the test dir

File: dbtmetabase/__init__.py
Comment (z3z1ma) [reply dbtmetabase/__init__.py:143]: No, I do generally think store_true & store_false is a little more readable / understandable compared to a flag that needs a string arg which evals to a bool but for the sake of consistency in expected function, I regressed this change.

File: dbtmetabase/__init__.py
Comment (z3z1ma) [reply dbtmetabase/__init__.py:18]: The default of 30 still lives at the innermost function call so if sync timeout is not specified, default time is still unchanged. The difference now is explicitly passing in the arg `sync_timeout` will cause fail fast behavior whereas not passing it in attempts sync for default of 30 seconds and continues execution regardless of exact match. This gives users finer control without effecting how they interface with the cli.

File: dbtmetabase/__init__.py
Comment (z3z1ma) [reply dbtmetabase/__init__.py:175]: Thanks for the catch. Regressed this change for the same reason as above. 
Want consistency in expected function. resolving comment

File: dbtmetabase/__init__.py
Comment (z3z1ma) [reply dbtmetabase/__init__.py]: I went ahead and normalized them all to not have a period as they were prior. If lib author has a pref change though we can go with that

File: dbtmetabase/metabase.py
Comment (z3z1ma) [reply dbtmetabase/metabase.py:22]: I went ahead and made this more explicit to reflect that it should be either bool or str

**Comment by z3z1ma** [comment]
@gouline 

I think the ball is in your court now. 

Note that I already address empty strings referred to by #23 
We only pass descriptions in 2 places, exporting columns and exporting tables. For tables, a simple logical check since without a description, there is nothing to propagate. And for columns, a simple ternary statement that checks for empty string.
I think in the future, @fernandobrito approach would be slick:
> It's a descriptor (https://realpython.com/python-descriptors/) to be used in class attributes that automatically convert empty strings to None.

but for now we are ok. 

The other issue raised in #24 I think should be examined and integrated if it addresses the issue after this PR so we dont step on our toes. Once thats done we can integrate changes easily I think.

File: dbtmetabase/parsers/dbt_folder.py
Comment (gouline) [inline dbtmetabase/parsers/dbt_folder.py:187]: Use f-strings, they're more readable.

File: dbtmetabase/parsers/dbt_folder.py
Comment (gouline) [inline dbtmetabase/parsers/dbt_folder.py:195]: Extract the comment out of the clause, long comments break the flow for the formatter. Put it before or after the assignment.

File: dbtmetabase/parsers/dbt_folder.py
Comment (gouline) [inline dbtmetabase/parsers/dbt_folder.py:181]: `"." not in mb_column.fk_target_table`

File: dbtmetabase/parsers/dbt_manifest.py
Comment (gouline) [inline dbtmetabase/parsers/dbt_manifest.py:44]: What are you joining here? Manifest path is already a string.

File: dbtmetabase/parsers/dbt_manifest.py
Comment (gouline) [inline dbtmetabase/parsers/dbt_manifest.py:108]: Redundant default `""` in both cases, since you're comparing with a literal.

File: dbtmetabase/parsers/dbt_folder.py
Comment (z3z1ma) [reply dbtmetabase/parsers/dbt_folder.py:181]: adjusted, thanks!

File: dbtmetabase/parsers/dbt_folder.py
Comment (z3z1ma) [reply dbtmetabase/parsers/dbt_folder.py:187]: f strings used for readability improvement

File: dbtmetabase/parsers/dbt_manifest.py
Comment (z3z1ma) [reply dbtmetabase/parsers/dbt_manifest.py:44]: removed os path join call

File: dbtmetabase/__init__.py
Comment (z3z1ma) [reply dbtmetabase/__init__.py]: yep removed those two for uniformity, thanks for the catch

File: dbtmetabase/__init__.py
Comment (z3z1ma) [reply dbtmetabase/__init__.py]: sounds good, agreed

File: README.rst
Comment (z3z1ma) [reply README.rst]: reverted

**Review by gouline — APPROVED** [review]
Looks good! Just need you to resolve a merge conflict from a minor PR merged in the meantime and we're good to go.

**Comment by gouline** [comment]
You renamed it in the wrong direction. Master has requirements-test.txt in the filename, setup and CI config.

**Comment by z3z1ma** [comment]
> You renamed it in the wrong direction. Master has requirements-test.txt in the filename, setup and CI config.

Thanks correcting now.

**Comment by gouline** [comment]
Looks like you're still editing. Let me know when it's ready for review.

**Comment by z3z1ma** [comment]
Its good to go @gouline, 

Freshly ran this branch on my production instances of Metabase/dbt too with both manifest and yml parser. Super easy. Good stuff. We are fully integrated into CD flow whenever we push any changes to dbt repo, our mb is synchronized. Onto posting my exposure parser PR once this is merged.

**Comment by gouline** [comment]
I see you pushed another commit after your response, but I'll assume that's it.

Thank you so much @z3z1ma and @fernandobrito for your efforts. I'm merging this and keeping it in master while we test it and merge other pull requests as they come along (there's one open now with a fix for an issue in master). Once that's all done, I will push a release to PyPI.

**Comment by fernandobrito** [comment]
Amazing work!

I will also try to find some time at work tomorrow to run what you have now on master on our production dbt, as so far I have been running my fork on our CD pipeline.

**Comment by fernandobrito** [comment]
I was able to swap my production setup from using my fork to using master, with the only exception that on master either `schema` or `schema_excludes` is required. I wrote about it here: https://github.com/gouline/dbt-metabase/pull/28/files#diff-399311141fa635ad8842613b2e8d964a08594d589911c6d1c0d5dc7d39bc44e9L72

**Pull request merged**
